### ISSUE TITLE
Redis Worker are not starting

### ISSUE THREAD

**Issue by nbungi** [ICB-DCM/pyABC#393]
Hi Yannick,

I am currently facing the following problem - I start the fmc/pyabc pipeline like usually but none of the workers starts to work and I also do not get an error message.
The output for all the output-files look ok to me:
```
INFO:History:Start <ABCSMC(id=1, start_time=2021-01-04 14:09:57.327789, end_time=None)>
INFO:ABC:t: 0, eps: 65.
```

```
....
INFO:Redis-Worker:Start redis worker. Max run time 428400.0s, HOST=h04b0302, PID=17930
INFO:Redis-Worker:Started subprocess with pid 17930
....
```
If I check with the abc-redis-manager info I get the following output
`Workers=0 Evaluations=0 Acceptances=0/256 (generation 0)`

and if I log onto one the nodes of the workers I see that no morpheus simulations have started.

The redis-server itself is working and is reachable (redis-cli ping).
`14775:M 04 Jan 2021 15:05:15.352 * Ready to accept connections`

Any idea what might be the problem?

**Comment by yannikschaelte** [comment]
This part of pyABC was recently [reworked](https://github.com/ICB-DCM/pyABC/commit/d77ef496f878be819358add04b92c1b81ba1c4dd#diff-a5f411c8cf2bafa275f0b87424b1eda396abf3817fa4d616ac63e858bf13496a), so that would look sus. The last message you get is the "Start redis worker ... PID=...". Directly after [that line](https://github.com/ICB-DCM/pyABC/blob/master/pyabc/sampler/redis_eps/cli.py#L90) pyABC creates the connection to the server (maybe we should swap those lines), so first guess would be that the connection cannot be established -- e.g. due to network issues (port+ip accessible from the worker system?). In that case the worker process should have stopped immediately and the error written to stderr/stdout.

**Comment by nbungi** [comment]
I also tried to ping the redis-server from the worker node itself which is working.

**Comment by yannikschaelte** [comment]
To clarify: Are the worker processes running idly, or are they terminated immediately?

**Comment by nbungi** [comment]
They are working idly. So the job is still running on the cluster.

**Comment by yannikschaelte** [comment]
Ok, then above I misunderstood. The output `Workers=0 Evaluations=0 Acceptances=0/256 (generation 0)` indicates that the server has at least been told that there is an analysis to work on, but somehow the workers don't get the note that they should start working on a population (?). Can you show the main script? I.e. ABCSMC+new+run?

**Comment by nbungi** [comment]
```
redis_sampler = RedisEvalParallelSampler(host=args.ip, port=args.port, password=password, look_ahead=True, look_ahead_delay_evaluation=True, log_file='look_ahead.csv')


abc = ABCSMC(model, prior, distance, summary_statistics=sum_stat, transitions=transition, 
             sampler=redis_sampler, population_size=pop_strategy,
             eps=eps)

# Database
db_name = str.split(filename, sep = '.')[0]
db = "sqlite:///" + os.path.join(cur_path, db_name + '.db') 

abc.new(db, r.observation("Exp"))

# Abbruchkriterium
history = abc.run(minimum_epsilon=5)
```

also tried to start the pipeline with #386 (redis_sampler.stop() ) but the workers did not start as well

**Comment by yannikschaelte** [comment]
Has the server been previously used for another analysis? But if so, in run() the server should [raise](https://github.com/ICB-DCM/pyABC/blob/master/pyabc/sampler/redis_eps/sampler.py#L70).

**Comment by nbungi** [comment]
The server had not been running for another analysis until today. I started the pipeline today around 1 pm and checked upon it around 1 hour later which is when I realized that no worker had started working and cancelled the jobs. I restarted a server with the same port to check if I could connect to the redis-server at around 3pm with the same port. However, it might well be that the redis-server job now runs on another node on the cluster so it probably has a different IP and thus probably would not raise "seems busy with another analysis". For the server would have to run on the same IP and Port to raise this error or?

**Comment by yannikschaelte** [comment]
Yes. "Server" = "Redis Server main process". If you start `redis-server` anew and make sure that you connect your abc analysis and the workers to the correct address, then all should work, so weird it doesn't ...

**Comment by nbungi** [comment]
Indeed. In the meantime, I also forwarded this issue to the cluster IT, maybe this problem is specific to the cluster I am using.

**Comment by yannikschaelte** [comment]
Probably makes sense. How I would further proceed: 1) Set up a server locally and check whether a mini-setup (popsize=2) works there. If this works, it must be the network connections. 2) Check whether look-ahead = False works. If this works, it must be look-ahead, which is new. This is rather unlikely though, as unit tests work. 3) Insert print() statements at random places in the code (redis_eps/sampler.py+work.py and inference/smc.py) to see how far the algorithm gets.

**Comment by nbungi** [comment]
W.r.t. 2) it did not make a difference. 
W.r.t 1) a local setup worked as expected. Therefore, it probably is a cluster related issue and I'll wait for the response of the IT.
Thank you for your time and help!

**Comment by yannikschaelte** [comment]
As it seems to be cluster-specific, you could next try whether all works when the workers are on the same node as the server. It could be related to permitted communication pathways.

**Comment by EmadAlamoudi** [comment]
I'm having the same issue on the Bonna Cluster, It seems that even thou the workers are connected successfully to the master when checking the run's info, I got:
`Workers=0 Evaluations=0 Acceptances=0/1000 (generation 0)`

**Comment by nbungi** [comment]
@EmadAlamoudi 
I am not 100% sure on what exactly problem is to be honest (wether it is the cluster itself or only the redis-server). But I got the pipeline working (at least for now) using the following approach. 
Try using the advice given here #386 if you used the same Port/Redis-server previously.
After initializing the redis-server I put a sleep timer for some seconds and then call `redis.stop()`

W.r.t. to the cluster there were still some problems on some nodes for me. If you realize the pipeline is not working you may try to exclude the nodes, on which the redis-server and the python-script were submitted but did not run. If your cluster uses SLURM, you can do that by using the exclude option `--exclude=NODE`. If you find a node on which the pipeline starts working you may try to use that node in the future specifically by using `--nodelist=NODE`. This may help if the problem is node specific and the redis/python script is always submitted on the same node.

Hope this works for you!

**Comment by EmadAlamoudi** [comment]
Thanks @nbungi for the fast reply! I have tried the two suggested solutions and none works for me. However, I have tried another one suggested by @FelipeR888 which exchanges the submitting of the python script by the submitting of the workers. Meaning we call create workers first and then we call the python script. That one works for me.

**Comment by yannikschaelte** [comment]
I think this should have been resolved in 0.10.14, as discussed separately. Let us know if issues remain!
